filebeat/input/filestream/internal/input-logfile/harvester.go

```diff
@@ -177,10 +204,18 @@ func (hg *defaultHarvesterGroup) Start(ctx input.Context, s Source) {
 
 		err = hg.harvester.Run(ctx, s, cursor, publisher)
 		if err != nil && err != context.Canceled {
+			hg.readers.remove(srcID)
 			return fmt.Errorf("error while running harvester: %v", err)
 		}
+		// If the context was not cancelled it means that the Harvester is stopping because of
+		// some internal decision, not due to outside interaction.
+		// If it is stopping itself, it must clean up the bookkeeper.
+		if ctx.Cancelation.Err() != context.Canceled {
+			hg.readers.remove(srcID)
+		}
```

Comment:
shouldn't we always remove the ID on shutdown?

Reply:
Yes. But I had to separate it into several cases because calling `hg.Stop`, `hg.readers.remove` and `hg.Start` after one another lead to random order of execution. In some cases it made handling truncation undefined, because first the `Prospector` stopped the `Harvester` for a given `Source`.  `Stop` is already removing the ID from the bookkeeper. When the new `Harvester` was starting for the truncated file, the previous `Harvester` might be still shutting down and as `hg.readers.remove` was deferred the newly started `Harvester` was cancelled by the previous one sometimes.

It is still not perfect. One open question what happens the context is cancelled for some reason not by hg.readers.remove. So far I haven't found a case that would cause problem with this method. Maybe when the `Prospector` is stopped I have to delete all running sources from the bookkeeper?

Reply:
I see. It seems to work for now. Still, we should try to clean this up in a follow up PR. Shutdown and group source ID handling is not fully intuitive and a potential source of bugs in the future.

Reply:
Ok. It definitely needs more work and we need a bit more time to come up with a clearer solution.

Reply:
Also, I hope when we introduce the `copytruncate` prospector most of the complexity due to supporting truncation in the default prospector can be moved and refactored. The complexity is coming from the facts that we cannot anticipate how the files are truncated and rotated. With specialized prospectors, we can handle special cases a bit better.